File: leetcode3/CountNodesEqualtoAverageofSubtree/a.cpp
/* Copyright (C) 2023 by iamslash */

#include <cstdio>
#include <array>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 9ms 68.48% 12.1MB 18.44%
// dfs
// O(N) O(lgN)
class Solution {
private:
  int best = 0;
  array<int, 2> dfs(TreeNode* u) {
    // base
    if (u == nullptr) {
      return {0, 0};
    }
    // recursion
    auto lrst = dfs(u->left);
    auto rrst = dfs(u->right);
    int sum = lrst[0] + rrst[0] + u->val;
    int cnt = lrst[1] + rrst[1] + 1;
    if (sum / cnt == u->val) {
      best++;
    }
    return {sum, cnt};
  }
public:
  int averageOfSubtree(TreeNode* root) {
    dfs(root);
    return best;
  }
};

int main() {
  return 0;
}

File: leetcode3/CountNodesEqualtoAverageofSubtree/a.go
/* Copyright (C) 2023 by iamslash */

package main

import "fmt"

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

// 0ms 100.00% 4.9MB 5.71%
// dfs
// O(N) O(lgN)
func dfs(u *TreeNode, best *int) []int {
    // base
    if (u == nil) {
        return []int{0, 0}
    }
    // recursion
    lrst := dfs(u.Left, best)
    rrst := dfs(u.Right, best)
    sum  := lrst[0] + rrst[0] + u.Val
    cnt  := lrst[1] + rrst[1] + 1
    if (sum / cnt) == u.Val {
        (*best) += 1
    }        
    return []int{sum, cnt}
}
func averageOfSubtree(root *TreeNode) int {
    best := 0
    dfs(root, &best)
    return best
}

func main() {
    fmt.Println("hello world")
}
